### PR TITLE
[Fix-15036] [API] Fix task definition edit doesn't work

### LINKED ISSUE #15036: [Bug] [API] Fix task definition edit doesn't work
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

Problem description: All versions have the problem that task definition changes do not take effect in the workflow and vice versa. There are several points as follows:

1、Modify the task definition without a predecessor task, but it is not available in the workflow.
2、Modify the task definition with a predecessor task, but it is not available in the workflow.
3、Switch the workflow version and modify the task definition, which is not available in the workflow.

### What you expected to happen

The modifications in the task definition should also take effect in the workflow

### How to reproduce

修改前

### Anything else

_No response_

### Version

3.1.x

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by sdhzwc**
fix #15036

Problem description: All versions have the problem that task definition changes do not take effect in the workflow and vice versa. There are several points as follows:
1. Modify the task definition without a predecessor task, but it is not available in the workflow.
2、Modify the task definition with a predecessor task, but it is not available in the workflow.
3、Switch the workflow version and modify the task definition, which is not available in the workflow.

Dynamic demonstration of before and after repair comparison:
Before Repair
修改前

二、After Repair
修复后

### CODE REVIEW

**Comment by EricGao888** [comment]
@sdhzwc Could u plz use English for PR descriptions?

**Comment by sdhzwc** [comment]
> @sdhzwc Could u plz use English for PR descriptions?

Sorry, my English is not very good

**Comment by ruanwenjun** [comment]
> > @sdhzwc Could u plz use English for PR descriptions?
> 
> Sorry, my English is not very good Chinese is my native language, so describing it more smoothly

Thanks for your contribution, I changed the content to English.

**Review by sdhzwc — COMMENTED** [review]
I have completed the modifications

**Comment by qingwli** [comment]
Please fix ut 

```
Error:  Failures: 
Error:    TaskDefinitionServiceImplTest.updateTaskDefinition:201 Expected org.apache.dolphinscheduler.api.exceptions.ServiceException to be thrown, but nothing was thrown.

**Comment by sdhzwc** [comment]
#15036

File: dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/ProcessTaskRelationMapper.xml [hunk 1]
```diff
@@ -240,4 +241,18 @@
...
+        select
+        <include refid="baseSql"/>
+        from t_ds_process_task_relation
+        WHERE process_definition_code in (
+                                            SELECT
+                                            process_definition_code
+                                            FROM
+                                            t_ds_process_task_relation
+                                            WHERE
+                                            post_task_code = #{taskCode}
```
Comment (SbloodyS) [inline dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/ProcessTaskRelationMapper.xml:255]: The relationship between task and process can be one-to-many.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java [hunk 2]
```diff
@@ -781,21 +786,53 @@ private TaskDefinitionLog updateTask(User loginUser, long projectCode, long task
...
             for (ProcessTaskRelation processTaskRelation : processTaskRelations) {
                 if (taskCode == processTaskRelation.getPreTaskCode()) {
                     processTaskRelation.setPreTaskVersion(version);
                 } else if (taskCode == processTaskRelation.getPostTaskCode()) {
                     processTaskRelation.setPostTaskVersion(version);
                 }
-                int count = processTaskRelationMapper.updateProcessTaskRelationTaskVersion(processTaskRelation);
-                if (count != 1) {
+                processTaskRelation.setProcessDefinitionVersion(processDefinitionVersion);
+                int update_2 = processTaskRelationMapper.updateProcessTaskRelationTaskVersion(processTaskRelation);
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java]:
```suggestion
                int updateProcessDefinitionVersionCount = processTaskRelationMapper.updateProcessTaskRelationTaskVersion(processTaskRelation);
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java [hunk 3]
```diff
@@ -781,21 +786,53 @@ private TaskDefinitionLog updateTask(User loginUser, long projectCode, long task
...
+                    putMsg(result, Status.CREATE_PROCESS_TASK_RELATION_LOG_ERROR);
+                    throw new ServiceException(Status.CREATE_PROCESS_TASK_RELATION_LOG_ERROR);
+                }
+            }
+            ProcessDefinition processDefinition = processDefinitionMapper.queryByCode(processDefinitionCode);
+            processDefinition.setVersion(processDefinitionVersion);
+            processDefinition.setUpdateTime(now);
+            processDefinition.setUserId(loginUser.getId());
+            // update process definition
+            int update_3 = processDefinitionMapper.updateById(processDefinition);
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java]:
```suggestion
            int updateProcessDefinitionCount = processDefinitionMapper.updateById(processDefinition);
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java [hunk 4]
```diff
@@ -781,21 +786,53 @@ private TaskDefinitionLog updateTask(User loginUser, long projectCode, long task
...
+            ProcessDefinitionLog processDefinitionLog = new ProcessDefinitionLog(processDefinition);
+            processDefinitionLog.setOperateTime(now);
+            processDefinitionLog.setId(null);
+            processDefinitionLog.setOperator(loginUser.getId());
+            int insert_3 = processDefinitionLogMapper.insert(processDefinitionLog);
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java]:
```suggestion
            int insertProcessDefinitionLogCount = processDefinitionLogMapper.insert(processDefinitionLog);
```

**Review by SbloodyS — CHANGES_REQUESTED** [review]

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java (hunk 3 above)
Comment (sdhzwc) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java]: The modification has been completed

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java (hunk 4 above)
Comment (sdhzwc) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java]: The modification has been completed

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java (hunk 2 above)
Comment (sdhzwc) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TaskDefinitionServiceImpl.java]: The modification has been completed

File: dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/ProcessTaskRelationMapper.xml (hunk 1 above)
Comment (SbloodyS) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/ProcessTaskRelationMapper.xml:255]: BTW, `post_task_code` is not unique, `post_task_code` and `post_task_version` is unique.

Comment (sdhzwc) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/ProcessTaskRelationMapper.xml:255]: The modification has been completed

**Review by SbloodyS — APPROVED** [review]
LGTM

**Review by davidzollo — APPROVED** [review]
+1
LGTM

**Pull request merged**